File: System_Design/OCP_Followed.cpp
// Description: This code demonstrates the Open Close Principle (OCP) being followed in a shopping cart system. The DataBasePersistence Class is an abstract class with a pure virtual function saveToDB, which is implemented by three different classes (MongoDBPersistence, SQLDBPersistence, FileDBPersistence) for saving cart details to different types of databases. This design allows us to add new database types without modifying existing code, adhering to the OCP.

#include <bits/stdc++.h>
using namespace std;

class Product{
    public:
        string name;
        double price;
        Product(string n, double p){
            name = n; price = p;
        }
        void getDetails(){
            cout <<"Product: "<<name<<", Price: $"<<price<<endl;
        }
};

class ShoppingCart{     //---------------------------> Single Class Single method
    public:
    vector<Product*> products;
        void addProduct(Product* p){
            products.push_back(p);
            cout <<"Added "<<p->name<<" to cart."<<endl;
        }
        void removeProduct(Product* p){
            products.erase(remove(products.begin(), products.end(), p), products.end());
            cout <<"Removed "<<p->name<<" from cart."<<endl;
        }
        void viewCart(){
            cout <<"Shopping Cart:"<<endl;
            for(auto p : products){
                p->getDetails();
            }
        }
        double calculateTotal(){
            double total = 0;
            for(auto p : products){
                total += p->price;
            }
            return total;
        }
};
class InvoicePrinter{
    ShoppingCart* shoppingCart;
    public:
        void printInvoice(ShoppingCart* shoppingCart){
            cout <<"Invoice:"<<endl;
            for(auto p : shoppingCart->products){
                p->getDetails();
            }
            cout <<"Total: $"<<shoppingCart->calculateTotal()<<endl;
        }
};
class DataBasePersistence{  // Abstract Class
    ShoppingCart* shoppingCart;
    public:
        virtual void saveToDB(ShoppingCart* shoppingCart)=0; // Pure Virtual Function
};

class MongoDBPersistence : public DataBasePersistence{
    public:
        void saveToDB(ShoppingCart* shoppingCart) override{
            // Code to save cart details to Mongo database
            cout <<"Saving cart details to MongoDB..."<<endl;
        }
};

class SQLDBPersistence : public DataBasePersistence{
    public:
        void saveToDB(ShoppingCart* shoppingCart) override{
            // Code to save cart details to MySQL database
            cout <<"Saving cart details to My SQL DataBase..."<<endl;
        }
};

class FileDBPersistence : public DataBasePersistence{
    public:
        void saveToDB(ShoppingCart* shoppingCart) override{
            // Code to save cart details to File Storage System
            cout <<"Saving cart details to File Storage System..."<<endl;
        }
};

int main(){
    Product* p1 = new Product("Laptop", 999.99);
    Product* p2 = new Product("Smartphone", 499.99);
    Product* p3 = new Product("Headphones", 199.99);
    ShoppingCart* cart = new ShoppingCart();
    cart->addProduct(p1);
    cart->addProduct(p2);
    cart->addProduct(p3);
    cart->viewCart();
    double totalPrice= cart->calculateTotal();
    cout<<"Total Price: $"<<totalPrice<<endl;
    InvoicePrinter* printer = new InvoicePrinter();
    printer->printInvoice(cart);
    DataBasePersistence* mongoDB = new MongoDBPersistence();
    DataBasePersistence* sqlDB = new SQLDBPersistence();
    DataBasePersistence* fileStorage = new FileDBPersistence();
    mongoDB->saveToDB(cart);
    sqlDB->saveToDB(cart);
    fileStorage->saveToDB(cart);

    return 0;
}

File: System_Design/JAVA_Design/OCP_Followed.java
import java.util.ArrayList;

class Product{
    String name;
    double price;
    public Product(String name, double price){
        this.name = name;
        this.price = price;
    }
    public void getDetails(){
        System.out.println("Product: " + name + ", Price: $" + price);
    }
}
class ShoppingCart{
    ArrayList<Product>products = new ArrayList<>();
    public void addProduct(Product product){
        products.add(product);
        System.out.println("Product added to cart: " + product.name);
    }
    public void removeProduct(Product product){
        products.remove(product);
        System.out.println("Product removed from cart: " + product.name);
    }
    public void viewCart(){
        System.out.println("Shopping Cart:");
        for(Product product : products){
            product.getDetails();
        }
    }
    public double calculateTotal(){
        double total=0d;
        for(Product p : products){
            total += p.price;
        }
        return total;
    }
}
class InvoicePrinter{
    public void printInvoice(ShoppingCart cart){
        System.out.println("Invoice:");
        for(Product p : cart.products){
            p.getDetails();
        }
        System.out.println("Total: $" + cart.calculateTotal());
    }
}
abstract class DatabaseManager{
    abstract void save(ShoppingCart cart);
}
class MongoDBManager extends DatabaseManager{
    @Override
    void save(ShoppingCart cart) {
        System.out.println("Saving shopping cart to MongoDB...");
        // Code to save cart to MongoDB
    }
}
class SQLDatabaseManager extends DatabaseManager{
    @Override
    void save(ShoppingCart cart) {
        System.out.println("Saving shopping cart to SQL Database...");
        // Code to save cart to SQL Database
    }
}
class FileSystemManager extends DatabaseManager{
    @Override
    void save(ShoppingCart cart) {
        System.out.println("Saving shopping cart to File System...");
        // Code to save cart to File System
    }
}

public class OCP_Followed {
    public static void main(String[] args){
        Product p1 = new Product("Laptop", 999.99);
        Product p2 = new Product("Smartphone", 499.99);
        Product p3 = new Product("Headphones", 199.99);
        ShoppingCart cart = new ShoppingCart();
        cart.addProduct(p1);
        cart.addProduct(p2);
        cart.addProduct(p3);
        cart.viewCart();
        double total = cart.calculateTotal();
        System.out.println("Total Price: $" + total);
        InvoicePrinter printer = new InvoicePrinter();
        printer.printInvoice(cart);
        DatabaseManager dbManager = new MongoDBManager();
        DatabaseManager sqlManager = new SQLDatabaseManager();
        DatabaseManager fileManager = new FileSystemManager();
        sqlManager.save(cart);
        fileManager.save(cart);
        dbManager.save(cart);
    }
}
